chromaui/chromatic-e2e
# Smart story names (path, filename, describes, test title)

Issue: #AP-4015

## What Changed


For the story names, we now use more than just the test title. Previously, we've been only using the Playwright/Cypress test title (a Playwright test with `test('some-title')` would have the story name `some-title`), which resulted in no hierarchy and possible overwriting since tests in different files with the same title would collide into one story name.

Now, we use the test title, then any containing `describe()`s as folders, then the filename as a folder (truncating the `.spec.*` or `.cy.*` stuff), then any directories between the filename and the tests directory as folders.

I'm trying to keep as similar to possible as how CSF 3 does [auto-titling](https://storybook.js.org/docs/configure/sidebar-and-urls#csf-30-auto-titles), so I'm not replacing `_` or `-` with spaces, or capitalizing things, or doing any modifications besides stripping off the extension (`.spec.ts|.cy.ts`) of the file name. I'm not doing the [redundancy-removing](https://storybook.js.org/docs/configure/sidebar-and-urls#csf-30-auto-titles) that CSF does, however (reducing `MyComponent/MyComponent.stories.ts` to `MyComponent`), since I bet it's unlikely that Playwright or Cypress users would folder their individual test files in folders of the same name. But I could be assuming too much there.  

### Example
Given the following Playwright test:
```ts
// tests/ is root tests directory
// this file path is tests/some-dir/file.spec.ts

test.describe('a containing describe', () => {
  test('test title', () => {
    // actual test stuff
  })
})
```

the resulting story name would be `some-dir/file/a containing describe/test title`.

### Note

This is a **breaking change** for folks, as the first build they run after installing this version of the package will result in all-new snapshots (since all the story names will have changed).

It also means we don't have to manually add `/` to our test titles to get things foldered nicely in Storybook!

## How to test


* Check out the `chromatic-e2e` repo
* In the Playwright and tests, add some `test.describe()` blocks around some `test()`s (`describe()`s around `it()`s for Cypress)
* In a test title, add a period to the beginning of the title (like `test('.someMethod() should behave')`)
* Run `yarn build && yarn test:playwright && yarn test:cypress`
* Run `yarn archive-storybook:playwright` and `yarn archive-storybook:cypress`
* Verify that the story names include the `describe` titles as folder names
* Verify that the tests' filenames are folders
* Verify that the test title with the period did not get truncated
* Create a folder at the tests level
* Move a tests file into this folder
* Build, test, and archive-storybook again
* Verify that this folder is included as a folder in Storybook for this test
